### PR TITLE
[core] Add source location to status

### PR DESCRIPTION
**Pull request by dentiny**
Current implementation for `ray::Status` is not good enough, it lacks the ability to pinpoint error location, which is also the reason why some prefer exception, which propagate source and stacktrace automatically.

This PR is the first step for status improvement, will followup with other PRs to make a few macros for error status propagation, which automatically fills in the `RAY_LOC`.
Example,
- `RAY_THROW_IF_ERROR`
- `RAY_RETURN_IF_ERROR`

### CODE REVIEW

File: src/ray/common/status.h
Comment (dayshah) [inline src/ray/common/status.h:137]: adding a constructor that takes an rvalue string too might be a small thing that's worth it or just forwarding string and defining constructor in header?

**Review by dayshah — APPROVED** [review]
nice

File: src/ray/common/status.h
Comment (dentiny) [reply src/ray/common/status.h:137]: I think the correct way is to take value type, with no lvaule/rvalue overload.
Checkout "modern effective C++" item-41 for details.

I could followup with another PR later.

File: src/ray/common/source_location.cc
Comment (jjyao) [inline src/ray/common/source_location.cc:25]: Will `os << loc.filename << ":" << loc.line_no` be better?

File: src/ray/common/source_location.h
Comment (dentiny) [reply src/ray/common/source_location.h]: Updated.

File: src/ray/common/source_location.cc
Comment (dentiny) [reply src/ray/common/source_location.cc]: SG, updated.

File: src/ray/common/source_location.h
Comment (jjyao) [inline src/ray/common/source_location.h:26]: won't we have dangling pointer issue?

Comment (dentiny) [reply src/ray/common/source_location.h:26]: No, string view is not POD.

Comment (dentiny) [reply src/ray/common/source_location.h:26]: The only place to access should be status class and ostream printer

Comment (jjyao) [reply src/ray/common/source_location.h:26]: Discussed offline, it's only safe because filename only points to `__FILE__` which won't be freed. We should mention that in the comment.

Comment (dentiny) [reply src/ray/common/source_location.h:26]: Sure, comment added.

**Review by jjyao — APPROVED** [review]

**Comment by dentiny** [comment]
Hi @jjyao , could you please help me enable automerge for this PR? I don't have the permission, thank you!

**Comment by aslonnie** [comment]
the test is failing to build on windows with error:

````
src/ray/common/source_location_test.cc(35): error C7555: use of designated initializers requires at least '/std:c++20'
````

cc @can-anyscale

File: src/ray/common/source_location.h
Comment (aslonnie) [inline src/ray/common/source_location.h:38]: so, you might need to add a constructor.

(unless you are in the mood of upgrading c++ compiling env / flags in windows..)

Comment (dentiny) [reply src/ray/common/source_location.h:38]: Ok, let me fix it. Also curious how did it skip through CI pipeline? I did see window test in pipeline

Comment (dentiny) [reply src/ray/common/source_location.h:38]: Fix PR: https://github.com/ray-project/ray/pull/48512
Sorry for the inconvenience!

Comment (aslonnie) [reply src/ray/common/source_location.h:38]: > Also curious how did it skip through CI pipeline?

windows tests do not run on premerge by default. it requires clicking a button.

**Comment by can-anyscale** [comment]
This is breaking src/ray/common/source_location_test and is blocking the release, I'm reverting to unblock.

**Pull request merged**
